File: backend/src/main/resources/static/scripts/utils/comparationAlgorithm.js
import levensteinDistance from "./levensteinAlgorithm.js";

// Adjust this threshold based on tests
export default function comparationAlgorithm(mainProduct, candidateProducts) {
    // Score
    const productsScore = {}

    // Calculate score for each product by ()
    for (const product of candidateProducts) {
        const productBrandFallback = product.brand || product.normalized_name.includes(mainProduct.brand) ? mainProduct.brand : ""
        const pid = `${product.product_id}@${product.supermarket_name}`
        let brandScore = levensteinDistance(mainProduct.brand, productBrandFallback) / mainProduct.brand.length
        let nameScore = levensteinDistance(mainProduct.productName, product.normalized_name) / mainProduct.productName.length
        let keywordsScore = keywordsDistance(mainProduct.keywords, product.keywords) / mainProduct.keywords.length

        // Calculate final score with weights
        productsScore[pid] = keywordsScore ** 4 + nameScore + brandScore
        if (keywordsScore > .8 || nameScore > .8 || brandScore > .8) productsScore[pid] = Infinity
    }

    // Get the best product by score
    candidateProducts.sort((a, b) => productsScore[`${a.product_id}@${a.supermarket_name}`] - productsScore[`${b.product_id}@${b.supermarket_name}`])

    const bestBySupermarket = {}
    candidateProducts.forEach(product => {
        const score = productsScore[`${product.product_id}@${product.supermarket_name}`]
        if (!bestBySupermarket[product.supermarket_name] && score < Infinity) {
            bestBySupermarket[product.supermarket_name] = product
        }
    })

    return bestBySupermarket
}

export function keywordsDistance(k1, k2) {
    let missing = 0
    for (const keyword of k1) {
        if (!k2.join(",").includes(keyword)) {
            missing++
        }
    }

    return missing
}

File: backend/src/main/java/es/daw01/savex/components/ComparationAlgorithm.java
package es.daw01.savex.components;

import org.springframework.stereotype.Component;

import es.daw01.savex.DTOs.products.ProductDTO;
import es.daw01.savex.utils.LevenshteinUtils;

import java.util.*;

@Component
public class ComparationAlgorithm {

    /**
     * Compare a main product with a list of candidate products and return the best product by supermarket
     * 
     * @param mainProduct Main product to compare with the list of candidate products
     * @param candidateProducts List of products to compare with the main product
     * @return bestBySupermarket
     */
    public Map<String, ProductDTO> compareProducts(ProductDTO mainProduct, List<ProductDTO> candidateProducts) {
        Map<String, Double> productsScore = new HashMap<>();

        for (ProductDTO product : candidateProducts) {
            String productBrandFallback = (product.brand() != null && !product.brand().isEmpty()) 
                ? product.brand()
                : (product.normalized_name().contains(mainProduct.brand()) ? mainProduct.brand() : "");

            // Calculate brand score
            double brandScore = LevenshteinUtils.calculateLevenshteinDistance(mainProduct.brand(), productBrandFallback);
            brandScore = brandScore / (double) mainProduct.brand().length();

            // Calculate name score
            double nameScore = LevenshteinUtils.calculateLevenshteinDistance(mainProduct.normalized_name(), product.normalized_name());
            nameScore = nameScore / (double) mainProduct.normalized_name().length();
            
            // Calculate keywords score
            double keywordsScore = calculateKeywordsDistance(mainProduct.keywords(), product.keywords());
            keywordsScore = keywordsScore / (double) mainProduct.keywords().size();

            double finalScore = keywordsScore + nameScore + brandScore;
            if (keywordsScore > 0.8 || nameScore > 0.8 || brandScore > 0.8) {
                finalScore = Double.POSITIVE_INFINITY;
            }

            productsScore.put(product.product_id(), finalScore);
        }

        // Sort candidate products by score
        candidateProducts.sort(Comparator.comparingDouble(product -> productsScore.get(product.product_id())));

        // Get the best product by supermarket
        return getBestProductBySupermarket(candidateProducts);
    }

    // Private methods -------------------------------------------------------->>

    private Map<String, ProductDTO> getBestProductBySupermarket(List<ProductDTO> candidateProducts) {
        Map<String, ProductDTO> bestBySupermarket = new HashMap<>();
        for (ProductDTO product : candidateProducts) {
            if (!bestBySupermarket.containsKey(product.supermarket_name())) {
                bestBySupermarket.put(product.supermarket_name(), product);
            }
        }
        
        return bestBySupermarket;
    }

    private double calculateKeywordsDistance(List<String> k1, List<String> k2) {
        int missing = 0;
        for (String keyword : k1) {
            if (!String.join(",", k2).contains(keyword)) {
                missing++;
            }
        }
        return missing;
    }
}